### PR TITLE
CH4/OFI:  Add Cray PMI support

### PR DESCRIPTION
**Pull request by zhenggb72**
Cray PMI is a mix of PMI ver 1 and ver 2. The new code to support Cray PMI
is now under macro USE_CRAYPMI_API.
New option 'craypmi' is added to configure support it: --enable-pmi=craypmi.

### CODE REVIEW

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 1]
```diff
@@ -14,6 +14,11 @@
 #include "ofi_impl.h"
 #include "mpir_cvars.h"
 #include "pmi.h"
+#ifdef USE_PMI2_API
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_init.h]: Does `USE_CRAYPMI_API` imply this macro as well? It is not set by your `subconfigure.m4` changes.

**Comment by zhenggb72** [comment]
USE_CRAYPMI_API does not imply that macro. I probably should not include this change in this patch. But this is rather a fix for using PMI2. However, I do not intend to fix all PMI2 problem in CH4 in this patch.

**Comment by raffenet** [comment]
@zhenggb72 so does Cray PMI include all functions (both pmi and pmi2) in `pmi.h`?

**Comment by zhenggb72** [comment]
Yes, Cray PMI only needs pmi.h.  pmi.h also includes "pim2.h".

File: src/pmi/craypmi/subconfigure.m4 [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+AM_COND_IF([BUILD_PMI_CRAY],[
+
+# sets CPPFLAGS and LDFLAGS
+PAC_SET_HEADER_LIB_PATH([pmi])
+
+AC_CHECK_HEADER([pmi.h], [], [AC_MSG_ERROR([could not find pmi.h.  Configure aborted])])
+AC_DEFINE(USE_CRAYPMI_API, 1, [Define if Cray PMI API must be used])
+AC_CHECK_LIB([pmi], [PMI_Init],
+             [
+              PAC_PREPEND_FLAG([$CRAY_PMI_POST_LINK_OPTS -lpmi], [WRAPPER_LIBS])
```
Comment (raffenet) [inline src/pmi/craypmi/subconfigure.m4]: Where do the `CRAY_PMI_POST_LINK_OPTS` come from? User env?

**Comment by raffenet** [comment]
Got it. This looks very close once we remove the PMI2 ifdefs. See also my second comment on the diff. Thanks for looking into this!

File: src/pmi/craypmi/subconfigure.m4 (hunk 2 above)
Comment (zhenggb72) [inline src/pmi/craypmi/subconfigure.m4]: yes, it is defined in module pmi in Cray environment.

**Comment by zhenggb72** [comment]
I pushed the change to remove the USE_PMI2_API in the first comment.

File: src/mpid/ch4/src/ch4_types.h [hunk 3]
```diff
@@ -14,7 +14,11 @@
 #include <mpidimpl.h>
 #include <stdio.h>
 #include "mpir_cvars.h"
+#ifdef USE_PMI2_API
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_types.h]: I can remove this one, too, right?

Comment (zhenggb72) [inline src/mpid/ch4/src/ch4_types.h]: Yes. However, no harm leaving there too. 
Can you please do that? I am leaving out of town today.

Comment (raffenet) [inline src/mpid/ch4/src/ch4_types.h]: Yup, I will take care of it and merge. Thanks.

**Pull request merged**